Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp

```diff
@@ -1969,21 +1969,143 @@ void WebsiteDataStore::resetStoragePersistedState(CompletionHandler<void()>&& co
-    return String();
+    return cacheDirectoryFileSystemRepresentation("CacheStorage"_s, baseCacheDirectory);
 }
 
-String WebsiteDataStore::defaultAlternativeServicesDirectory(const String&)
+String WebsiteDataStore::defaultGeneralStorageDirectory(const String& baseDataDirectory)
 {
-    // FIXME: Implement.
-    return String();
+#if PLATFORM(PLAYSTATION) || USE(GLIB)
+    return websiteDataDirectoryFileSystemRepresentation("storage"_s, baseDataDirectory);
+#else
+    return websiteDataDirectoryFileSystemRepresentation("Storage"_s, baseDataDirectory);
+#endif
+}
+
+String WebsiteDataStore::defaultNetworkCacheDirectory(const String& baseCacheDirectory)
+{
+#if PLATFORM(PLAYSTATION) || USE(GLIB)
+    return cacheDirectoryFileSystemRepresentation("WebKitCache"_s, baseCacheDirectory);
+#else
+    return cacheDirectoryFileSystemRepresentation("NetworkCache"_s, baseCacheDirectory);
+#endif
+}
+
+String WebsiteDataStore::defaultApplicationCacheDirectory(const String& baseCacheDirectory)
+{
+#if PLATFORM(PLAYSTATION) || USE(GLIB)
+    return cacheDirectoryFileSystemRepresentation("applications"_s, baseCacheDirectory);
+#else
+    return cacheDirectoryFileSystemRepresentation("ApplicationCache"_s, baseCacheDirectory);
+#endif
+}
+
+String WebsiteDataStore::defaultMediaCacheDirectory(const String& baseCacheDirectory)
+{
+    return cacheDirectoryFileSystemRepresentation("MediaCache"_s, baseCacheDirectory);
+}
+
+String WebsiteDataStore::defaultIndexedDBDatabaseDirectory(const String& baseDataDirectory)
+{
+#if PLATFORM(PLAYSTATION)
+    return websiteDataDirectoryFileSystemRepresentation("indexeddb"_s, baseDataDirectory);
+#elif USE(GLIB)
+    return websiteDataDirectoryFileSystemRepresentation(String::fromUTF8("databases" G_DIR_SEPARATOR_S "indexeddb"), baseDataDirectory);
+#else
+    return websiteDataDirectoryFileSystemRepresentation("IndexedDB"_s, baseDataDirectory);
+#endif
+}
+
+String WebsiteDataStore::defaultServiceWorkerRegistrationDirectory(const String& baseDataDirectory)
+{
+#if PLATFORM(PLAYSTATION) || USE(GLIB)
+    return websiteDataDirectoryFileSystemRepresentation("serviceworkers"_s, baseDataDirectory);
+#else
+    return websiteDataDirectoryFileSystemRepresentation("ServiceWorkers"_s, baseDataDirectory);
+#endif
+}
+
+String WebsiteDataStore::defaultWebSQLDatabaseDirectory(const String& baseDataDirectory)
+{
+#if PLATFORM(PLAYSTATION)
+    return websiteDataDirectoryFileSystemRepresentation("websql"_s, baseDataDirectory);
+#elif USE(GLIB)
+    return websiteDataDirectoryFileSystemRepresentation("databases"_s, baseDataDirectory);
+#else
+    return websiteDataDirectoryFileSystemRepresentation("WebSQL"_s, baseDataDirectory);
+#endif
+}
+
+String WebsiteDataStore::defaultHSTSStorageDirectory(const String& baseCacheDirectory)
+{
+#if USE(GLIB)
+    return websiteDataDirectoryFileSystemRepresentation(""_s, baseCacheDirectory);
```

Comment:
> I know this is existing behavior but it looks a bit dangerous, as I think we usually check if directory is empty to decide if we a type can be stored. So it seems this should either be websiteDataDirectoryFileSystemRepresentation("HSTS"_s, baseDataDirectory), if you want to create a HSTS storage, or emptyString() (unless you do want to store this under the root cache directory...)

The code is working properly: the HSTS database really does go in the root cache directory. This might look weird, but it's suitable for GLib because the HSTS database is exactly one file, and creating a directory to store only one file is kinda overkill.

Ideally the HSTS database would be partitioned by your OriginStorageManager to prevent HSTS supercookie abuse. Well, perhaps _all_ website data should be partitioned. But anyway, for now it's just one file.

> And it seems weird that this calls websiteDataDirectoryFileSystemRepresentation with baseCacheDirectory...

Oh good catch. That's a preexisting bug, dating all the way back to https://commits.webkit.org/214897@main. O_O But it only happens if the application does not configure a base cache directory, so it doesn't affect Epiphany or MiniBrowser. Probably does affect most other applications, though. Drat.

I'll have to think about this. I'll fix it for the new GLib API for sure. For the old API, if I fix it, then the old HSTS database at the old location will be leaked on disk. That doesn't seem great, but I don't really want to write a migrator, so maybe I will leave the old API version untouched. Hmm....

Reply:
I decided it's easiest to fix it only for the new API.

Reply:
BTW if you recommend moving it under a directory instead of putting it in the base cache directory, I suppose now is the best time to do that. I don't think it matters too much either way.

Reply:
> I know this is existing behavior but it looks a bit dangerous, as I think we usually check if directory is empty to decide if we a type can be stored. So it seems this should either be websiteDataDirectoryFileSystemRepresentation("HSTS"_s, baseDataDirectory), if you want to create a HSTS storage, or emptyString() (unless you do want to store this under the root cache directory...)

I decided to move it in the new API version.